File: java/src/test/java/io/github/marchliu/jse/QueryTest.java
package io.github.marchliu.jse;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static io.github.marchliu.jse.Sql.QUERY_FIELDS;

class QueryTest {

    @Test
    void basicQuery() {
        Map<String, Object> query = Map.of(
                "$expr", List.of("$pattern", "$*", "author of", "$*")
        );

        Engine engine = new Engine(new Env.ExpressionEnv());
        Object result = engine.execute(query);
        String sql = (String) result;

        assertTrue(sql.contains("select"));
        assertTrue(sql.contains("subject, predicate, object, meta"));
        assertTrue(sql.contains("from statement as s"));
        assertTrue(sql.contains("author of"));
        assertTrue(sql.contains("triple"));
        assertTrue(sql.contains("offset 0"));
        assertTrue(sql.contains("limit 100"));
    }

    @Test
    void combinedQuery() {
        Map<String, Object> query = Map.of(
                "$query", List.of(
                        "$and",
                        List.of(
                                List.of("$pattern", "Liu Xin", "author of", "$*"),
                                List.of("$pattern", "$*", "author of", "$*")
                        )
                )
        );

        Engine engine = new Engine(new Env.ExpressionEnv());
        String result = (String) engine.execute(query);

        assertTrue(result.contains("select " + QUERY_FIELDS));
        assertTrue(result.contains("from statement"));
        assertTrue(result.contains("Liu Xin"));
        assertTrue(result.contains("author of"));
        assertTrue(result.contains(" and "));
        assertTrue(result.contains("offset 0"));
        assertTrue(result.contains("limit 100"));
    }
}

File: typescript/src/__tests__/query.test.ts
import { describe, it, expect } from "vitest";
import { Engine, Env, UTILS_FUNCTORS, SQL_FUNCTORS } from "../index.js";

function createEngine() {
  const env = new Env();
  env.load(UTILS_FUNCTORS, SQL_FUNCTORS);
  return new Engine(env);
}

describe("query expressions", () => {
  const engine = createEngine();

  it("basic $expr $pattern query", () => {
    const query = {
      $expr: ["$pattern", "$*", "author of", "$*"],
    };

    const result = engine.execute(query) as string;

    expect(result).toContain("select");
    expect(result).toContain("subject, predicate, object, meta");
    expect(result).toContain("from statement as s");
    expect(result).toContain("author of");
    expect(result).toContain("triple");
    expect(result).toContain("offset 0");
    expect(result).toContain("limit 100");
  });

  it("combined $query with $and patterns", () => {
    const query = {
      $query: [
        "$and",
        [
          ["$pattern", "Liu Xin", "author of", "$*"],
          ["$pattern", "$*", "author of", "$*"],
        ],
      ],
    };

    const result = engine.execute(query) as string;

    expect(result).toContain("select subject, predicate, object, meta");
    expect(result).toContain("from statement");
    expect(result).toContain("Liu Xin");
    expect(result).toContain("author of");
    expect(result).toContain(" and ");
    expect(result).toContain("offset 0");
    expect(result).toContain("limit 100");
  });
});
